File: python/src/bleepstore/storage/gcp.py
"""Google Cloud Storage gateway backend for BleepStore.

Proxies all data operations to an upstream GCS bucket via gcloud-aio-storage.
Metadata stays in local SQLite — this backend handles raw bytes only.

Key mapping:
    Objects:  {prefix}{bleepstore_bucket}/{key}
    Parts:    {prefix}.parts/{upload_id}/{part_number}

Credentials are resolved via GCS Application Default Credentials
(GOOGLE_APPLICATION_CREDENTIALS, gcloud auth, metadata server).
"""

import hashlib
import logging
from collections.abc import AsyncIterator

from gcloud.aio.storage import Storage

logger = logging.getLogger(__name__)

# Streaming chunk size: 64 KB (matches local and AWS backends)
_CHUNK_SIZE = 64 * 1024

# GCS compose() supports at most 32 source objects per call.
_MAX_COMPOSE_SOURCES = 32


class GCPGatewayBackend:
    """Storage backend that proxies to a Google Cloud Storage bucket.

    All BleepStore buckets/objects are stored under a single upstream
    GCS bucket with a key prefix to namespace them.

    Attributes:
        bucket_name: The upstream GCS bucket name.
        project: The GCP project ID.
        prefix: Key prefix for all objects in the upstream bucket.
    """

    def __init__(
        self,
        bucket_name: str,
        project: str = "",
        prefix: str = "",
    ) -> None:
        self.bucket_name = bucket_name
        self.project = project
        self.prefix = prefix
        self._client: Storage | None = None

    def _gcs_name(self, bucket: str, key: str) -> str:
        """Map a BleepStore bucket/key to an upstream GCS object name."""
        return f"{self.prefix}{bucket}/{key}"

    def _part_name(self, upload_id: str, part_number: int) -> str:
        """Map a multipart part to an upstream GCS object name."""
        return f"{self.prefix}.parts/{upload_id}/{part_number}"

    async def init(self) -> None:
        """Create the gcloud-aio-storage client and verify the upstream bucket exists.

        Raises:
            ValueError: If the upstream bucket does not exist or is inaccessible.
        """
        self._client = Storage()

        # Verify bucket exists by attempting to list with maxResults=1
        try:
            await self._client.list_objects(
                self.bucket_name,
                params={"maxResults": "1"},
            )
        except Exception as e:
            await self._client.close()
            self._client = None
            raise ValueError(
                f"Cannot access upstream GCS bucket '{self.bucket_name}': {e}"
            ) from e

        logger.info(
            "GCP gateway backend initialized: bucket=%s project=%s prefix='%s'",
            self.bucket_name,
            self.project,
            self.prefix,
        )

    async def close(self) -> None:
        """Close the gcloud-aio-storage client session."""
        if self._client is not None:
            await self._client.close()
            self._client = None

    async def put(self, bucket: str, key: str, data: bytes) -> str:
        """Upload an object to the upstream GCS bucket.

        Computes MD5 locally for a consistent ETag.

        Returns:
            The hex-encoded MD5 of the stored data.
        """
        gcs_name = self._gcs_name(bucket, key)
        md5 = hashlib.md5(data).hexdigest()

        await self._client.upload(
            self.bucket_name,
            gcs_name,
            data,
        )
        return md5

    async def get(self, bucket: str, key: str) -> bytes:
        """Download an object from the upstream GCS bucket.

        Raises:
            FileNotFoundError: If the object does not exist.
        """
        gcs_name = self._gcs_name(bucket, key)
        try:
            return await self._client.download(self.bucket_name, gcs_name)
        except Exception as e:
            if _is_not_found(e):
                raise FileNotFoundError(
                    f"Object not found: {bucket}/{key}"
                ) from e
            raise

    async def get_stream(
        self, bucket: str, key: str, offset: int = 0, length: int | None = None
    ) -> AsyncIterator[bytes]:
        """Stream an object from the upstream GCS bucket in 64KB chunks.

        Raises:
            FileNotFoundError: If the object does not exist.
        """
        gcs_name = self._gcs_name(bucket, key)

        headers: dict[str, str] = {}
        if offset > 0 or length is not None:
            if length is not None:
                end = offset + length - 1
                headers["Range"] = f"bytes={offset}-{end}"
            else:
                headers["Range"] = f"bytes={offset}-"

        try:
            stream_resp = await self._client.download_stream(
                self.bucket_name,
                gcs_name,
                headers=headers if headers else None,
            )
        except Exception as e:
            if _is_not_found(e):
                raise FileNotFoundError(
                    f"Object not found: {bucket}/{key}"
                ) from e
            raise

        while True:
            chunk = await stream_resp.read(_CHUNK_SIZE)
            if not chunk:
                break
            yield chunk

    async def delete(self, bucket: str, key: str) -> None:
        """Delete an object from the upstream GCS bucket.

        Idempotent — catches 404 silently (GCS errors on delete of
        non-existent objects unlike S3).
        """
        gcs_name = self._gcs_name(bucket, key)
        try:
            await self._client.delete(self.bucket_name, gcs_name)
        except Exception as e:
            if _is_not_found(e):
                return  # Idempotent: treat as success
            raise

    async def exists(self, bucket: str, key: str) -> bool:
        """Check if an object exists in the upstream GCS bucket.

        Uses a Range: bytes=0-0 download to avoid fetching the full object.
        """
        gcs_name = self._gcs_name(bucket, key)
        try:
            await self._client.download(
                self.bucket_name,
                gcs_name,
                headers={"Range": "bytes=0-0"},
            )
            return True
        except Exception as e:
            if _is_not_found(e):
                return False
            raise

    async def put_part(
        self, bucket: str, key: str, upload_id: str, part_number: int, data: bytes
    ) -> str:
        """Store a multipart upload part as a temporary GCS object.

        Parts are stored at {prefix}.parts/{upload_id}/{part_number}.
        Computes MD5 locally for a consistent ETag.

        Returns:
            The hex-encoded MD5 of the part data.
        """
        part_name = self._part_name(upload_id, part_number)
        md5 = hashlib.md5(data).hexdigest()

        await self._client.upload(
            self.bucket_name,
            part_name,
            data,
        )
        return md5

    async def assemble_parts(
        self,
        bucket: str,
        key: str,
        upload_id: str,
        part_numbers: list[int],
    ) -> str:
        """Assemble uploaded parts into the final object using GCS compose.

        GCS compose() supports at most 32 source objects per call. For >32
        parts, chains compose in batches of 32: compose each batch into an
        intermediate object, then compose the intermediates, repeating until
        a single object remains.

        Returns:
            The hex-encoded MD5 of the assembled object (best-effort).
        """
        final_name = self._gcs_name(bucket, key)
        source_names = [self._part_name(upload_id, pn) for pn in part_numbers]

        if len(source_names) <= _MAX_COMPOSE_SOURCES:
            # Simple case: single compose call
            await self._client.compose(
                self.bucket_name, final_name, source_names
            )
        else:
            # Chain compose in batches of 32
            intermediates = await self._chain_compose(source_names, final_name)
            # Clean up intermediate composite objects
            for name in intermediates:
                try:
                    await self._client.delete(self.bucket_name, name)
                except Exception:
                    logger.warning("Failed to clean up intermediate: %s", name)

        # Compute MD5 of the final assembled object
        data = await self._client.download(self.bucket_name, final_name)
        return hashlib.md5(data).hexdigest()

    async def _chain_compose(
        self, source_names: list[str], final_name: str
    ) -> list[str]:
        """Chain GCS compose calls for >32 sources.

        Returns a list of intermediate object names that should be cleaned up.
        """
        all_intermediates: list[str] = []
        current_sources = source_names

        generation = 0
        while len(current_sources) > _MAX_COMPOSE_SOURCES:
            next_sources: list[str] = []
            for i in range(0, len(current_sources), _MAX_COMPOSE_SOURCES):
                batch = current_sources[i : i + _MAX_COMPOSE_SOURCES]
                if len(batch) == 1:
                    # Single source — no compose needed, just pass through
                    next_sources.append(batch[0])
                    continue
                intermediate_name = f"{final_name}.__compose_tmp_{generation}_{i}"
                await self._client.compose(
                    self.bucket_name, intermediate_name, batch
                )
                next_sources.append(intermediate_name)
                all_intermediates.append(intermediate_name)
            current_sources = next_sources
            generation += 1

        # Final compose
        await self._client.compose(
            self.bucket_name, final_name, current_sources
        )
        return all_intermediates

    async def delete_parts(self, bucket: str, key: str, upload_id: str) -> None:
        """Delete all temporary part objects for a multipart upload.

        Lists objects under .parts/{upload_id}/ and deletes each one.
        """
        prefix = f"{self.prefix}.parts/{upload_id}/"

        resp = await self._client.list_objects(
            self.bucket_name,
            params={"prefix": prefix},
        )

        items = resp.get("items", [])
        for item in items:
            obj_name = item["name"]
            try:
                await self._client.delete(self.bucket_name, obj_name)
            except Exception as e:
                if not _is_not_found(e):
                    raise

    async def copy_object(
        self,
        src_bucket: str,
        src_key: str,
        dst_bucket: str,
        dst_key: str,
    ) -> str:
        """Copy an object using GCS server-side copy.

        Returns:
            The hex-encoded MD5 ETag of the copied object.
        """
        src_gcs_name = self._gcs_name(src_bucket, src_key)
        dst_gcs_name = self._gcs_name(dst_bucket, dst_key)

        await self._client.copy(
            self.bucket_name,
            src_gcs_name,
            self.bucket_name,
            new_name=dst_gcs_name,
        )

        # Compute MD5 by downloading the result
        data = await self._client.download(self.bucket_name, dst_gcs_name)
        return hashlib.md5(data).hexdigest()


def _is_not_found(exc: Exception) -> bool:
    """Check if an exception represents a 404 Not Found from GCS."""
    # gcloud-aio-storage raises aiohttp.ClientResponseError for HTTP errors
    status = getattr(exc, "status", None)
    if status == 404:
        return True
    # Also check for wrapped exceptions
    if hasattr(exc, "args") and exc.args:
        msg = str(exc.args[0]).lower()
        if "404" in msg or "not found" in msg:
            return True
    return False

File: golang/internal/storage/gcp.go
// Package storage provides the GCP Cloud Storage gateway backend for BleepStore.
//
// The GCP gateway backend proxies all data operations to an upstream GCS
// bucket via the official Go Cloud Storage client library. Metadata stays
// in local SQLite -- this backend handles raw bytes only.
//
// Key mapping:
//
//	Objects:  {prefix}{bleepstore_bucket}/{key}
//	Parts:    {prefix}.parts/{upload_id}/{part_number}
//
// Credentials are resolved via Application Default Credentials
// (GOOGLE_APPLICATION_CREDENTIALS, gcloud auth, metadata server).
package storage

import (
	"bytes"
	"context"
	"crypto/md5"
	"errors"
	"fmt"
	"io"
	"log"
	"strings"

	gcs "cloud.google.com/go/storage"
	"google.golang.org/api/iterator"
)

// maxComposeSources is the GCS limit on the number of source objects per
// Compose call.
const maxComposeSources = 32

// GCSAPI defines the subset of the GCS client interface that the gateway
// backend uses. This allows mocking in tests.
type GCSAPI interface {
	// NewWriter returns a writer for the given GCS object.
	NewWriter(ctx context.Context, bucket, object string) GCSWriter
	// NewReader returns a reader for the given GCS object.
	NewReader(ctx context.Context, bucket, object string) (io.ReadCloser, error)
	// Delete deletes the given GCS object.
	Delete(ctx context.Context, bucket, object string) error
	// Attrs returns the attributes of the given GCS object.
	Attrs(ctx context.Context, bucket, object string) (*GCSAttrs, error)
	// Copy copies a GCS object from src to dst within the same bucket.
	Copy(ctx context.Context, bucket, srcObject, dstObject string) (*GCSAttrs, error)
	// Compose composes multiple GCS source objects into a single destination object.
	Compose(ctx context.Context, bucket, dstObject string, srcObjects []string) (*GCSAttrs, error)
	// ListObjects lists objects with the given prefix.
	ListObjects(ctx context.Context, bucket, prefix string) ([]string, error)
}

// GCSWriter is a writer interface for writing to GCS objects.
type GCSWriter interface {
	io.WriteCloser
}

// GCSAttrs holds object attributes returned from GCS operations.
type GCSAttrs struct {
	Size int64
	MD5  []byte // raw MD5 hash bytes
}

// realGCSClient wraps the official GCS client to satisfy GCSAPI.
type realGCSClient struct {
	client *gcs.Client
}

func (c *realGCSClient) NewWriter(ctx context.Context, bucket, object string) GCSWriter {
	return c.client.Bucket(bucket).Object(object).NewWriter(ctx)
}

func (c *realGCSClient) NewReader(ctx context.Context, bucket, object string) (io.ReadCloser, error) {
	return c.client.Bucket(bucket).Object(object).NewReader(ctx)
}

func (c *realGCSClient) Delete(ctx context.Context, bucket, object string) error {
	return c.client.Bucket(bucket).Object(object).Delete(ctx)
}

func (c *realGCSClient) Attrs(ctx context.Context, bucket, object string) (*GCSAttrs, error) {
	attrs, err := c.client.Bucket(bucket).Object(object).Attrs(ctx)
	if err != nil {
		return nil, err
	}
	return &GCSAttrs{
		Size: attrs.Size,
		MD5:  attrs.MD5,
	}, nil
}

func (c *realGCSClient) Copy(ctx context.Context, bucket, srcObject, dstObject string) (*GCSAttrs, error) {
	src := c.client.Bucket(bucket).Object(srcObject)
	dst := c.client.Bucket(bucket).Object(dstObject)
	attrs, err := dst.CopierFrom(src).Run(ctx)
	if err != nil {
		return nil, err
	}
	return &GCSAttrs{
		Size: attrs.Size,
		MD5:  attrs.MD5,
	}, nil
}

func (c *realGCSClient) Compose(ctx context.Context, bucket, dstObject string, srcObjects []string) (*GCSAttrs, error) {
	dst := c.client.Bucket(bucket).Object(dstObject)
	var srcs []*gcs.ObjectHandle
	for _, name := range srcObjects {
		srcs = append(srcs, c.client.Bucket(bucket).Object(name))
	}
	attrs, err := dst.ComposerFrom(srcs...).Run(ctx)
	if err != nil {
		return nil, err
	}
	return &GCSAttrs{
		Size: attrs.Size,
		MD5:  attrs.MD5,
	}, nil
}

func (c *realGCSClient) ListObjects(ctx context.Context, bucket, prefix string) ([]string, error) {
	it := c.client.Bucket(bucket).Objects(ctx, &gcs.Query{Prefix: prefix})
	var names []string
	for {
		attrs, err := it.Next()
		if errors.Is(err, iterator.Done) {
			break
		}
		if err != nil {
			return nil, err
		}
		names = append(names, attrs.Name)
	}
	return names, nil
}

// GCPGatewayBackend implements the StorageBackend interface by proxying
// storage operations to Google Cloud Storage. This allows BleepStore to
// act as an S3-compatible gateway in front of GCS.
//
// All BleepStore buckets/objects are stored under a single upstream GCS bucket
// with a key prefix to namespace them.
type GCPGatewayBackend struct {
	// Bucket is the upstream GCS bucket name.
	Bucket string
	// Project is the GCP project ID.
	Project string
	// Prefix is the key prefix for all objects in the upstream bucket.
	Prefix string
	// client is the GCS client (satisfying GCSAPI interface).
	client GCSAPI
}

// NewGCPGatewayBackend creates a new GCPGatewayBackend configured to proxy
// to the specified GCS bucket. It initializes the GCS client using
// Application Default Credentials.
func NewGCPGatewayBackend(ctx context.Context, bucket, project, prefix string) (*GCPGatewayBackend, error) {
	client, err := gcs.NewClient(ctx)
	if err != nil {
		return nil, fmt.Errorf("creating GCS client: %w", err)
	}

	b := &GCPGatewayBackend{
		Bucket:  bucket,
		Project: project,
		Prefix:  prefix,
		client:  &realGCSClient{client: client},
	}

	// Verify the upstream bucket is accessible by listing with a small limit.
	_, err = b.client.ListObjects(ctx, bucket, "\x00nonexistent\x00")
	if err != nil {
		return nil, fmt.Errorf("cannot access upstream GCS bucket %q: %w", bucket, err)
	}

	log.Printf("GCP gateway backend initialized: bucket=%s project=%s prefix=%q", bucket, project, prefix)
	return b, nil
}

// NewGCPGatewayBackendWithClient creates a GCPGatewayBackend with a
// pre-configured GCS client. This is primarily used for testing with mock
// clients.
func NewGCPGatewayBackendWithClient(bucket, project, prefix string, client GCSAPI) *GCPGatewayBackend {
	return &GCPGatewayBackend{
		Bucket:  bucket,
		Project: project,
		Prefix:  prefix,
		client:  client,
	}
}

// gcsKey maps a BleepStore bucket/key to an upstream GCS object name.
func (b *GCPGatewayBackend) gcsKey(bucket, key string) string {
	return b.Prefix + bucket + "/" + key
}

// partKey maps a multipart part to an upstream GCS object name.
func (b *GCPGatewayBackend) partKey(uploadID string, partNumber int) string {
	return fmt.Sprintf("%s.parts/%s/%d", b.Prefix, uploadID, partNumber)
}

// PutObject uploads object data to the upstream GCS bucket. It reads all data,
// computes MD5 locally for a consistent ETag, then uploads to GCS.
func (b *GCPGatewayBackend) PutObject(ctx context.Context, bucket, key string, reader io.Reader, size int64) (int64, string, error) {
	gcsName := b.gcsKey(bucket, key)

	// Read all data to compute MD5 locally. GCS may return different ETags
	// or no MD5 for composite objects, so we compute our own.
	data, err := io.ReadAll(reader)
	if err != nil {
		return 0, "", fmt.Errorf("reading object data: %w", err)
	}

	h := md5.New()
	h.Write(data)
	etag := fmt.Sprintf(`"%x"`, h.Sum(nil))

	w := b.client.NewWriter(ctx, b.Bucket, gcsName)
	if _, err := io.Copy(w, bytes.NewReader(data)); err != nil {
		_ = w.Close()
		return 0, "", fmt.Errorf("uploading to GCS: %w", err)
	}
	if err := w.Close(); err != nil {
		return 0, "", fmt.Errorf("finalizing GCS upload: %w", err)
	}

	return int64(len(data)), etag, nil
}

// GetObject retrieves object data from the upstream GCS bucket.
// Returns the data stream, the object size, and an empty ETag (metadata store
// holds the authoritative ETag). The caller is responsible for closing the
// returned ReadCloser.
func (b *GCPGatewayBackend) GetObject(ctx context.Context, bucket, key string) (io.ReadCloser, int64, string, error) {
	gcsName := b.gcsKey(bucket, key)

	// Get attributes first for size.
	attrs, err := b.client.Attrs(ctx, b.Bucket, gcsName)
	if err != nil {
		if isGCSNotFound(err) {
			return nil, 0, "", fmt.Errorf("object not found: %s/%s", bucket, key)
		}
		return nil, 0, "", fmt.Errorf("getting object attrs from GCS: %w", err)
	}

	reader, err := b.client.NewReader(ctx, b.Bucket, gcsName)
	if err != nil {
		if isGCSNotFound(err) {
			return nil, 0, "", fmt.Errorf("object not found: %s/%s", bucket, key)
		}
		return nil, 0, "", fmt.Errorf("getting object from GCS: %w", err)
	}

	return reader, attrs.Size, "", nil
}

// DeleteObject removes an object from the upstream GCS bucket.
// Idempotent: catches 404 silently (GCS errors on delete of non-existent
// objects unlike S3).
func (b *GCPGatewayBackend) DeleteObject(ctx context.Context, bucket, key string) error {
	gcsName := b.gcsKey(bucket, key)

	err := b.client.Delete(ctx, b.Bucket, gcsName)
	if err != nil {
		if isGCSNotFound(err) {
			return nil // Idempotent: treat as success
		}
		return fmt.Errorf("deleting object from GCS: %w", err)
	}
	return nil
}

// CopyObject copies an object within the upstream GCS bucket using GCS
// server-side copy. Downloads the result to compute MD5 for a consistent ETag.
func (b *GCPGatewayBackend) CopyObject(ctx context.Context, srcBucket, srcKey, dstBucket, dstKey string) (string, error) {
	srcGCSName := b.gcsKey(srcBucket, srcKey)
	dstGCSName := b.gcsKey(dstBucket, dstKey)

	_, err := b.client.Copy(ctx, b.Bucket, srcGCSName, dstGCSName)
	if err != nil {
		if isGCSNotFound(err) {
			return "", fmt.Errorf("source object not found: %s/%s", srcBucket, srcKey)
		}
		return "", fmt.Errorf("copying object in GCS: %w", err)
	}

	// Download the copied object to compute MD5 for consistent ETag.
	reader, err := b.client.NewReader(ctx, b.Bucket, dstGCSName)
	if err != nil {
		return "", fmt.Errorf("reading copied object for ETag: %w", err)
	}
	data, err := io.ReadAll(reader)
	reader.Close()
	if err != nil {
		return "", fmt.Errorf("reading copied object data: %w", err)
	}

	h := md5.New()
	h.Write(data)
	etag := fmt.Sprintf(`"%x"`, h.Sum(nil))
	return etag, nil
}

// PutPart stores a multipart upload part as a temporary GCS object.
// Parts are stored at {prefix}.parts/{upload_id}/{part_number}.
// Computes MD5 locally for a consistent ETag.
func (b *GCPGatewayBackend) PutPart(ctx context.Context, bucket, key, uploadID string, partNumber int, reader io.Reader, size int64) (string, error) {
	pk := b.partKey(uploadID, partNumber)

	// Read all data to compute MD5 locally.
	data, err := io.ReadAll(reader)
	if err != nil {
		return "", fmt.Errorf("reading part data: %w", err)
	}

	h := md5.New()
	h.Write(data)
	etag := fmt.Sprintf(`"%x"`, h.Sum(nil))

	w := b.client.NewWriter(ctx, b.Bucket, pk)
	if _, err := io.Copy(w, bytes.NewReader(data)); err != nil {
		_ = w.Close()
		return "", fmt.Errorf("uploading part to GCS: %w", err)
	}
	if err := w.Close(); err != nil {
		return "", fmt.Errorf("finalizing part upload to GCS: %w", err)
	}

	return etag, nil
}

// AssembleParts composes the specified parts into a single GCS object using
// GCS Compose. GCS compose supports at most 32 source objects per call.
// For >32 parts, chains compose in batches of 32: compose each batch into
// an intermediate object, then compose the intermediates, repeating until
// a single object remains.
//
// Returns the composite ETag computed by downloading the final object.
func (b *GCPGatewayBackend) AssembleParts(ctx context.Context, bucket, key, uploadID string, partNumbers []int) (string, error) {
	finalName := b.gcsKey(bucket, key)
	sourceNames := make([]string, len(partNumbers))
	for i, pn := range partNumbers {
		sourceNames[i] = b.partKey(uploadID, pn)
	}

	if len(sourceNames) <= maxComposeSources {
		// Simple case: single compose call.
		_, err := b.client.Compose(ctx, b.Bucket, finalName, sourceNames)
		if err != nil {
			return "", fmt.Errorf("composing parts in GCS: %w", err)
		}
	} else {
		// Chain compose in batches of 32.
		intermediates, err := b.chainCompose(ctx, sourceNames, finalName)
		if err != nil {
			return "", err
		}
		// Clean up intermediate composite objects.
		for _, name := range intermediates {
			if delErr := b.client.Delete(ctx, b.Bucket, name); delErr != nil {
				log.Printf("Warning: failed to clean up intermediate: %s: %v", name, delErr)
			}
		}
	}

	// Compute MD5 of the final assembled object by downloading it.
	reader, err := b.client.NewReader(ctx, b.Bucket, finalName)
	if err != nil {
		return "", fmt.Errorf("reading assembled object for ETag: %w", err)
	}
	data, err := io.ReadAll(reader)
	reader.Close()
	if err != nil {
		return "", fmt.Errorf("reading assembled object data: %w", err)
	}

	h := md5.New()
	h.Write(data)
	etag := fmt.Sprintf(`"%x"`, h.Sum(nil))
	return etag, nil
}

// chainCompose chains GCS compose calls for >32 sources.
// Returns a list of intermediate object names that should be cleaned up.
func (b *GCPGatewayBackend) chainCompose(ctx context.Context, sourceNames []string, finalName string) ([]string, error) {
	var allIntermediates []string
	currentSources := sourceNames

	generation := 0
	for len(currentSources) > maxComposeSources {
		var nextSources []string
		for i := 0; i < len(currentSources); i += maxComposeSources {
			end := i + maxComposeSources
			if end > len(currentSources) {
				end = len(currentSources)
			}
			batch := currentSources[i:end]
			if len(batch) == 1 {
				// Single source: no compose needed, just pass through.
				nextSources = append(nextSources, batch[0])
				continue
			}
			intermediateName := fmt.Sprintf("%s.__compose_tmp_%d_%d", finalName, generation, i)
			_, err := b.client.Compose(ctx, b.Bucket, intermediateName, batch)
			if err != nil {
				return allIntermediates, fmt.Errorf("composing intermediate batch (gen=%d, offset=%d): %w", generation, i, err)
			}
			nextSources = append(nextSources, intermediateName)
			allIntermediates = append(allIntermediates, intermediateName)
		}
		currentSources = nextSources
		generation++
	}

	// Final compose.
	_, err := b.client.Compose(ctx, b.Bucket, finalName, currentSources)
	if err != nil {
		return allIntermediates, fmt.Errorf("final compose in GCS: %w", err)
	}
	return allIntermediates, nil
}

// DeleteParts removes all temporary part objects for a multipart upload.
// Lists objects under .parts/{upload_id}/ and deletes each one.
func (b *GCPGatewayBackend) DeleteParts(ctx context.Context, bucket, key, uploadID string) error {
	prefix := b.Prefix + ".parts/" + uploadID + "/"

	names, err := b.client.ListObjects(ctx, b.Bucket, prefix)
	if err != nil {
		return fmt.Errorf("listing parts for upload %s: %w", uploadID, err)
	}

	for _, name := range names {
		if delErr := b.client.Delete(ctx, b.Bucket, name); delErr != nil {
			if !isGCSNotFound(delErr) {
				return fmt.Errorf("deleting part %s: %w", name, delErr)
			}
		}
	}

	return nil
}

// CreateBucket is a no-op for the GCP gateway backend. All BleepStore buckets
// share a single upstream GCS bucket with key prefixes, so there is nothing
// to create on the GCS side.
func (b *GCPGatewayBackend) CreateBucket(ctx context.Context, bucket string) error {
	return nil
}

// DeleteBucket is a no-op for the GCP gateway backend. Bucket data is scoped
// by key prefix, so there is nothing to delete on the GCS side. The metadata
// store handles the actual bucket record deletion.
func (b *GCPGatewayBackend) DeleteBucket(ctx context.Context, bucket string) error {
	return nil
}

// ObjectExists checks whether an object exists in the upstream GCS bucket.
func (b *GCPGatewayBackend) ObjectExists(ctx context.Context, bucket, key string) (bool, error) {
	gcsName := b.gcsKey(bucket, key)

	_, err := b.client.Attrs(ctx, b.Bucket, gcsName)
	if err != nil {
		if isGCSNotFound(err) {
			return false, nil
		}
		return false, fmt.Errorf("checking object existence in GCS: %w", err)
	}
	return true, nil
}

// isGCSNotFound checks if a GCS error is a 404/not-found error.
func isGCSNotFound(err error) bool {
	if errors.Is(err, gcs.ErrObjectNotExist) {
		return true
	}
	if errors.Is(err, gcs.ErrBucketNotExist) {
		return true
	}
	// Check error message as fallback.
	if err != nil {
		msg := strings.ToLower(err.Error())
		if strings.Contains(msg, "not found") || strings.Contains(msg, "404") {
			return true
		}
	}
	return false
}

// Ensure GCPGatewayBackend implements StorageBackend at compile time.
var _ StorageBackend = (*GCPGatewayBackend)(nil)
